### PR TITLE
Automatically create tick formatters for str and callable inputs.

### PR DESCRIPTION
**Pull request by toddrjen**
## PR Summary

This allows `matplotlib.axis.Axis.set_major_formatter` and `matplotlib.axis.Axis.set_minor_formatter` to accept a ``str`` or function (really any ``callable``) in addition to a `matplotlib.ticker.Formatter`.  This will create and use a `matplotlib.ticker.StrMethodFormatter` and `matplotlib.ticker.FuncFormatter`, respectively.  This significantly reduces the barrier to using custom formatters in simple cases.

This was previously [discussed](https://mail.python.org/pipermail/matplotlib-devel/2020-March/001326.html) on the matplotlib-devel mailing list.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
Thanks @toddrjen !

To try and make sure feature PRs don't fall through the cracks, we starting to assign champions who are responsible for making sure the review process goes through, @story645 has agreed to champion this PR.

**Comment by toddrjen** [comment]
@tacaswell @story645 Great, thank you!

I don't know why the test coverage has dropped, I added new tests that, as far as I can tell, cover the new code paths.

Assuming this is a good approach, the big outstanding issue in my mind is whether this should be used more in the documentation.  There are several examples using the old-style `FormatStrFormatter` that could be switched to using this approach.

**Comment by story645** [comment]
@toddrjen can you please provide some examples here cause I'm struggling a bit in finding 'em in the PR

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1529,37 +1529,68 @@ def set_major_formatter(self, formatter):
         """
         Set the formatter of the major ticker.
 
+        In addition to a `~matplotlib.ticker.Formatter` instance,
```
Comment (anntzer) [inline lib/matplotlib/axis.py]: As argued in the mailing list (not repro'ing the discussion here) I'd like to argue in favor of
```
This accepts either a `.Formatter` instance, or a callable, which is
automatically wrapped in a `.FuncFormatter`.  To use a format string
(e.g. "{x:.4f}"), pass the corresponding bound string method
(`"{x:.4f}".format`).
```
which I don't actually think is harder to understand (in one case the user needs to learn about StrMethodFormatter, which is a highly matplotlib specific concept; in the other case, about bound methods, which is also somewhat advanced, but at least much more widely applicable).

(Then you can remove all the mentions of StrMethodFormatter.)

**Comment by toddrjen** [comment]
@story645  One example for the `str` approach used in the documentation is:

```Python
axs[5].xaxis.set_major_formatter('{x} km')
```

Which would replace:

```Python
from matplotlib import ticker
axs[5].xaxis.set_major_formatter(ticker.StrMethodFormatter('{x} km'))
```

One for the function approach used in the documentation:

```Python
axs[3].xaxis.set_major_formatter(lambda x, pos: str(x-5))
```

Which would replace:

```Python
from matplotlib import ticker
axs[3].xaxis.set_major_formatter(ticker.FuncFormatter(lambda x, pos: str(x-5)))
```

**Comment by jklymak** [comment]
In the pr it’s not clear that you need a pos argument to your function or why.

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py]: In my opinion the `StrMethodFormatter` is the most basic formatter.  There is probably a reason the documentation routinely uses the `FormatStrFormatter` rather than the `FuncFormatter`.  Further, in the documentation the string-related formatter are usually described first.  Learning to work with strings is one of the first thing new python developers encounter, while passing around function objects is much more advanced.

So although `StrMethodFormatter` is a new function, it isn't a new concept.  Once a beginning user sees they can use strings, they won't have to worry about functions in most cases.  Using a bound method like this, on the other hand, is a new concept for many users, one they wouldn't necessarily have to deal with otherwise.  This goes against the purpose of this change, which is to make things simpler and easier to use.

That is similar to the reason I intentionally don't even mention callables.  A beginning users isn't going to know what that is and is just going to get confused, while a more advanced user is going to understand that callables can be used wherever functions are normally used.

So although this makes the code simpler, it only really saves two lines.  And I think this comes at the expense of making things more difficult for beginning users, exactly the sort of people I was trying to help with this change.

**Comment by toddrjen** [comment]
@jklymak Where, exactly, do you think this should be mentioned?  Can you comment on the specific lines or blocks where you think this should be changed?  In most cases I am just using or following up on the existing description for the `FuncFormatter`.

**Comment by jklymak** [comment]
> @jklymak Where, exactly, do you think this should be mentioned? Can you comment on the specific lines or blocks where you think this should be changed? In most cases I am just using or following up on the existing description for the `FuncFormatter`.

It needs to be documented somewhere, otherwise how are people to know what a valid callable is?

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (jklymak) [reply lib/matplotlib/axis.py]: I agree with @toddrjen here.  We want to make this less arcane, not more arcane, and telling folks to pass the `format` method of a string is pretty arcane.   Most people want to just do: `format_my_ticks('%01.3f')`.  

WRT this, you apparently need the `x` in `'{x}'`?   This also needs to be documented?  If you are going to just refer people to the underlying Formatter classes to understand the syntax needed, then you lose a lot of the simplification here.

Comment (anntzer) [reply lib/matplotlib/axis.py]: I don't feel strongly enough to block over this, but enough to not approve :shrug:

**Comment by toddrjen** [comment]
@jklymak Please see the updated version.  I have added more explicit documentation.

**Comment by toddrjen** [comment]
I have added a new commit where I modified a bunch of existing examples and tutorials, and the previous documentation, to make better use of the new syntax.  These changes may be more controversial so I added them to a separate commit so they would be easier to roll back.  

Please take a look and see what you think.

I don't know why the tests are failing.  It doesn't appear to be related to my commit but I am not sure.

**Comment by toddrjen** [comment]
@jklymak All tests are passing now.  Is there anything else I need to do or is this ready?

**Comment by toddrjen** [comment]
@story645  All tests are passing now. Is there anything else I need to do or is this ready?

**Comment by story645** [comment]
Sorry I owe still you feedback on this, just wanted to note that I'm +1 on this idea since learning that spottily is basically using this sort of thing as their [motivating example](https://labs.spotify.com/2018/11/15/introducing-chartify-easier-chart-creation-in-python-for-data-scientists/) for writing a  new charting library.

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
...
+        this also accepts a ``str`` or function.
+
+        For a ``str`` a `~matplotlib.ticker.StrMethodFormatter` is used.
+        The field used for the value must be labeled ``'x'`` and the field used
```
Comment (story645) [inline lib/matplotlib/axis.py:1522]: I understand you're using field to be more general, but I'd go a drop more technical here 'cause I think it can be clearer:
``` 
the tick value argument must be named ``'x'`` and the tick position argument must be labeled ``'pos'``
```

File: lib/matplotlib/axis.py [hunk 3]
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
...
+        for the position must be labeled ``'pos'``.
+        See the  `~matplotlib.ticker.StrMethodFormatter` documentation for
+        more information.
+
+        For a function a `~matplotlib.ticker.FuncFormatter` is used.
```
Comment (story645) [inline lib/matplotlib/axis.py:1527]: For a function, a

File: lib/matplotlib/axis.py [hunk 4]
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
...
+        The function should take in two inputs (a tick value ``x`` and a
```
Comment (story645) [inline lib/matplotlib/axis.py:1528]: must, not should? also two arguments (or two fields if you use the terminology mentioned above - I guess above all keep to consistent terminology)

File: lib/matplotlib/axis.py [hunk 5]
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
...
+        position ``pos``), and return a string containing the corresponding
+        tick label.
```
Comment (story645) [inline lib/matplotlib/axis.py:1556]: I think this level of detail should only exist in one place, especially since you cross link to the canonical source.

File: lib/matplotlib/axis.py [hunk 6]
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
...
+        See the  `~matplotlib.ticker.FuncFormatter` documentation for
+        more information.
```
Comment (story645) [inline lib/matplotlib/axis.py:1584]: Same as above, it should be in one canonical place with cross references. Maybe a short hand in comments for devs.

File: lib/matplotlib/axis.py [hunk 7]
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
...
+        # Don't allow any other TickHelper to avoid easy-to-make errors,
+        # like using a Locator instead of a Formatter.
+        elif (callable(formatter) and
+              not isinstance(formatter, mticker.TickHelper)):
+            formatter = mticker.FuncFormatter(formatter)
+        else:
+            cbook._check_isinstance(mticker.Formatter, formatter=formatter)
+
         if (isinstance(formatter, mticker.FixedFormatter)
                 and len(formatter.seq) > 0
```
Comment (story645) [inline lib/matplotlib/axis.py:1602]: this s may have been in the code before, but why does it having a length matter for this check?

File: lib/matplotlib/axis.py [hunk 8]
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
...
-                and not isinstance(self.minor.locator, mticker.FixedLocator)):
+                and not isinstance(level.locator, mticker.FixedLocator)):
             cbook._warn_external('FixedFormatter should only be used together '
                                  'with FixedLocator')
-        self.isDefault_minfmt = False
-        self.minor.formatter = formatter
+
+        if level == self.major:
+            self.isDefault_majfmt = False
```
Comment (story645) [inline lib/matplotlib/axis.py:1608]: really unclear what this attributes mean/do so not sure if the code is doing the right thing here

File: lib/matplotlib/tests/test_axes.py [hunk 9]
```diff
@@ -204,6 +204,28 @@ def test_formatter_ticker():
     ax.autoscale_view()
 
 
+def test_auto_formatter():
```
Comment (story645) [inline lib/matplotlib/tests/test_axes.py:207]: can you break this into two tests, one for the string functionality and one for the function functionality?

Comment (toddrjen) [reply lib/matplotlib/tests/test_axes.py:207]: Done

File: lib/matplotlib/axis.py (hunk 8 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py:1608]: This attribute is being used by the original, individual functions.  I just combined them.  But I added a test to make sure they are being set properly.

File: lib/matplotlib/axis.py (hunk 7 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py:1602]: I don't really understand FixedFormatter and wouldn't want to touch it.

File: lib/matplotlib/axis.py (hunk 5 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py:1556]: I was specifically told by someone else that I needed to add more detail here.

File: lib/matplotlib/axis.py (hunk 6 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py:1584]: Again, I was told I add more detail here.

File: lib/matplotlib/axis.py (hunk 4 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py:1528]: Fixed

File: lib/matplotlib/axis.py (hunk 3 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py:1527]: Okay

File: doc/users/next_whats_new/2020-03-06-auto-tick-formatters.rst [hunk 10]
```diff
@@ -0,0 +1,7 @@
+Allow tick formatters to be set with str or function inputs
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply doc/users/next_whats_new/2020-03-06-auto-tick-formatters.rst:1]: They can already be set with strings and functions, just indirectly.  I want to make it clear that this is directly as inputs.  And this is an optional approach, which isn't clear from your version.  So I don't think your version is as clear about what is actually changing here.

File: doc/users/next_whats_new/2020-03-06-auto-tick-formatters.rst [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+------------------------------------------------------------------------
+`~.Axis.set_major_formatter` and `~.Axis.set_minor_formatter`
+now accept `str` or function inputs in addition to `~.ticker.Formatter`
+instances. For a `str` a `~.ticker.StrMethodFormatter` is automatically
+created and used. For a function a `~.ticker.FuncFormatter` is automatically
+created and used.
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply doc/users/next_whats_new/2020-03-06-auto-tick-formatters.rst:7]: Fixed.

File: examples/mplot3d/surface3d.py [hunk 12]
```diff
@@ -31,9 +31,27 @@
...
+# """"""""""
+#
+# The use of the following functions, methods, classes and modules is shown
+# in this example:
+
+import matplotlib
+matplotlib.pyplot.subplots
+matplotlib.axis.Axis.set_major_formatter
+matplotlib.axis.Axis.set_major_locator
+matplotlib.ticker.StrMethodFormatter
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply examples/mplot3d/surface3d.py:57]: Added.

File: examples/pyplots/dollar_ticks.py [hunk 13]
```diff
@@ -7,16 +7,15 @@
...
 
 fig, ax = plt.subplots()
 ax.plot(100*np.random.rand(20))
 
-formatter = ticker.FormatStrFormatter('$%1.2f')
-ax.yaxis.set_major_formatter(formatter)
+# Use automatic StrMethodFormatter
+ax.yaxis.set_major_formatter('${x:1.2f}')
 
 for tick in ax.yaxis.get_major_ticks():
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply examples/pyplots/dollar_ticks.py:20]: Okay, I changed it.

File: examples/showcase/anatomy.py [hunk 14]
```diff
@@ -24,13 +24,14 @@
...
-    return "%.2f" % x
+    return f"{x:.2f}"
 
 ax.xaxis.set_major_locator(MultipleLocator(1.000))
 ax.xaxis.set_minor_locator(AutoMinorLocator(4))
 ax.yaxis.set_major_locator(MultipleLocator(1.000))
 ax.yaxis.set_minor_locator(AutoMinorLocator(4))
-ax.xaxis.set_minor_formatter(FuncFormatter(minor_tick))
+# FuncFormatter is created and used automatically
+ax.xaxis.set_minor_formatter(minor_tick)
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply examples/showcase/anatomy.py:34]: The "new machinery" is solely in the `set_minor_formatter` and `set_major_formatter` methods.  The existing `FuncFormatter` behavior has not been changed.  Everything that works before will still work.

File: tutorials/introductory/lifecycle.py [hunk 15]
```diff
@@ -187,9 +187,14 @@
 
 ###############################################################################
 # For labels, we can specify custom formatting guidelines in the form of
-# functions by using the :class:`ticker.FuncFormatter` class. Below we'll
-# define a function that takes an integer as input, and returns a string
-# as an output.
+# functions. Below we'll define a function that takes an integer as input, and
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply tutorials/introductory/lifecycle.py:190]: Okay I changed this.

File: lib/matplotlib/axis.py (hunk 2 above)
Comment (toddrjen) [reply lib/matplotlib/axis.py:1522]: "Field" is the term used by the upstream Python documentation, so I think we should stick with that.  It really isn't an "argument".

File: lib/matplotlib/ticker.py [hunk 16]
```diff
@@ -415,8 +421,8 @@ class StrMethodFormatter(Formatter):
     """
     Use a new-style format string (as used by `str.format`) to format the tick.
 
-    The field used for the value must be labeled *x* and the field used
-    for the position must be labeled *pos*.
+    The field used for the tick value must be labeled *x* and the field used
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply lib/matplotlib/ticker.py:424]: Again, this is the term used by the upstream Python documentation.  It isn't really an argument, so I wouldn't want to use that term.

File: lib/matplotlib/ticker.py [hunk 17]
```diff
@@ -159,6 +159,12 @@
   ax.yaxis.set_major_formatter(ymajor_formatter)
   ax.yaxis.set_minor_formatter(yminor_formatter)
 
+In addition to `.Formatter`, `~.Axis.set_major_formatter` and
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply lib/matplotlib/ticker.py:162]: Done.

File: tutorials/text/text_intro.py [hunk 18]
```diff
@@ -350,27 +349,29 @@
 
 ##############################################################################
 #  Finally, we can specify functions for the formatter using
-# `matplotlib.ticker.FuncFormatter`.
+# `matplotlib.ticker.FuncFormatter`.  Further, like
+# `matplotlib.ticker.StrMethodFormatter`, simply passing a function will
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply tutorials/text/text_intro.py:353]: Done.

File: tutorials/text/text_intro.py [hunk 19]
```diff
@@ -285,16 +285,16 @@
 # Instead of making a list of all the tickalbels, we could have
 # used `matplotlib.ticker.StrMethodFormatter` (new-style ``str.format()``
 # format string) or `matplotlib.ticker.FormatStrFormatter` (old-style '%'
-# format string) and passed it to the ``ax.xaxis``.
+# format string) and passed it to the ``ax.xaxis``.  A
+# `matplotlib.ticker.StrMethodFormatter` can also be created by simply
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply tutorials/text/text_intro.py:289]: Done

File: examples/ticks_and_spines/custom_ticker1.py [hunk 20]
```diff
@@ -21,13 +20,26 @@
...
-formatter = FuncFormatter(millions)
+    return '${:1.1f}M'.format(x*1e-6)
 
 fig, ax = plt.subplots()
-ax.yaxis.set_major_formatter(formatter)
-plt.bar(x, money)
-plt.xticks(x, ('Bill', 'Fred', 'Mary', 'Sue'))
+# Use automatic FuncFormatter creation
+ax.yaxis.set_major_formatter(millions)
+ax.bar(x, money)
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply examples/ticks_and_spines/custom_ticker1.py:28]: Changed this.

File: examples/ticks_and_spines/tick-formatters.py [hunk 21]
```diff
@@ -34,46 +34,87 @@ def setup(ax, title):
...
-# Otherwise, one cannot be sure where the labels will end up.
-positions = [0, 1, 2, 3, 4, 5]
-labels = ['A', 'B', 'C', 'D', 'E', 'F']
-axs[1].xaxis.set_major_locator(ticker.FixedLocator(positions))
-axs[1].xaxis.set_major_formatter(ticker.FixedFormatter(labels))
+# A function can also be used directly as a formatter. The function must take
+# two arguments: ``x`` for the tick value and ``pos`` for the tick position,
+# and must return a ``str``  This creates a FuncFormatter automatically.
+setup(axs[1], title="lambda x, pos: str(x-5)")
+axs[1].xaxis.set_major_formatter(lambda x, pos: str(x-5))
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply examples/ticks_and_spines/tick-formatters.py:50]: Okay.

File: tutorials/intermediate/artists.py [hunk 22]
```diff
@@ -652,16 +652,13 @@ class in the matplotlib API, and the one you will be working with most
...
 np.random.seed(19680801)
 
 fig, ax = plt.subplots()
 ax.plot(100*np.random.rand(20))
 
-formatter = ticker.FormatStrFormatter('$%1.2f')
-ax.yaxis.set_major_formatter(formatter)
+ax.yaxis.set_major_formatter('${x:1.2f}')
 
 for tick in ax.yaxis.get_major_ticks():
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply tutorials/intermediate/artists.py:663]: I'll need to check on this.

[the remark this answers is not in the document]
Comment (toddrjen) [reply tutorials/intermediate/artists.py:663]: It appears to be possible but I need to do some work to get it working.

File: lib/matplotlib/ticker.py [hunk 23]
```diff
@@ -159,6 +159,12 @@
...
+`~.Axis.set_minor_formatter` also accept a ``str`` or function.  In the case of
+a ``str``, a `.StrMethodFormatter` will be automatically created and applied
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply lib/matplotlib/ticker.py:164]: Okay, done.

File: examples/ticks_and_spines/tick-formatters.py [hunk 24]
```diff
@@ -34,46 +34,87 @@ def setup(ax, title):
             fontsize=14, fontname='Monospace', color='tab:blue')
 
 
-fig, axs = plt.subplots(7, 1, figsize=(8, 6))
+fig, axs = plt.subplots(9, 1, figsize=(8, 8))
 
-# Null formatter
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply examples/ticks_and_spines/tick-formatters.py:39]: Okay, done.

File: tutorials/introductory/lifecycle.py [hunk 25]
```diff
@@ -187,9 +187,14 @@
...
+# returns a string as an output. When used with `.Axis.set_major_formatter` or
+# `.Axis.set_minor_formatter`, they will automatically create and use a
+# :class:`ticker.FuncFormatter` class.
+#
+# For this function, the ``x`` argument is the original tick label and ``pos``
+# is the tick position.  We will only use ``x`` here but both arguments are
+# needed.
```
[the remark this answers is not in the document]
Comment (toddrjen) [reply tutorials/introductory/lifecycle.py:197]: There are two ways I can think of to "magic" this away.  First I can try using two arguments, if there is a TypeError I can try one argument, then fail.  Alternatively I can calculate the number of arguments.  The first approach is fast but may silence valid errors.  The second is slower but more reliable.  Which would you prefer?

File: lib/matplotlib/axis.py (hunk 5 above)
Comment (story645) [reply lib/matplotlib/axis.py:1562]: Ok, so we discussed this on the call and consensus is that the canonical explanation should be in `set_major_formatter`  and the other places should cross link to it. `_set_formatter` doesn't need the full documentation since it is a private method. The cost of having to maintain 3 copies of the same docs outweighs the potential clarity gains since the docs can be cross reference.

Comment (toddrjen) [reply lib/matplotlib/axis.py:1562]: Okay, I did this.  Although it won't help people as much when using `help` directly.

**Comment by toddrjen** [comment]
@story645 All the discussed changes should be implemented.  Please take a look.  I am still waiting on feedback on a few issues.

File: tutorials/introductory/lifecycle.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (story645) [reply tutorials/introductory/lifecycle.py:197]: what's the time cost on trying to compute the errors?

**Comment by story645** [comment]
Thanks for your patience, been swamped with Passover. @jklymak and @anntzer have your concerns been addressed?

**Comment by anntzer** [comment]
As stated above I'm basically fine with the PR, but don't like one very specific point (direct support for strings) so don't particularly want to approve it with that feature.

File: .flake8 [hunk 26]
```diff
@@ -180,6 +180,7 @@ per-file-ignores =
     examples/misc/svg_filter_line.py: E402
     examples/misc/svg_filter_pie.py: E402
     examples/misc/table_demo.py: E201
+    examples/mplot3d/surface3d.py: E402
```
Comment (jklymak) [inline .flake8:183]: This looks unrelated to the PR?  Maybe something wrong w a rebase?

**Comment by toddrjen** [comment]
@jklymak 
> Do we want to replace _every_ instance of `StrMethodFormatter` and `FuncFormatter` by this idiom? There is some advantage to pointing out the ticker module and the existence of the suite of Formatters.

I didn't replace every instance.  `examples/ticks_and_spines/tick-formatters.py`  still explains how to use the existing method.

File: tutorials/introductory/lifecycle.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (toddrjen) [reply tutorials/introductory/lifecycle.py:197]: On the order of tens of microseconds.  Using a try...except is on the order of microseconds.

**Comment by story645** [comment]
Hi Todd, thanks for hanging in there! We're aiming to get this into 3.3 (rc out in a week or so).  

Giving it some more thought, keeping the current formatter signature seems fine as it's consistent with the docs and what folks will find on stackoverflows. Can revisit in a later PR. 

Do you feel it's ready to merge or do you have any additions to make?

**Comment by story645** [comment]
Also now thinking the arguments fix should actually be on FuncFormatter since it's not clear why it needs two https://matplotlib.org/_modules/matplotlib/ticker.html#FuncFormatter

**Comment by toddrjen** [comment]
@story645 If all the tests pass I think it is ready to merge.
